core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```
@@ -196,19 +227,93 @@ public void writeTo(StreamOutput out) throws IOException {
         this.parentChildType = parentChildType;
         return this;
     }
 
+    public InnerHitBuilder setParentChildType(String parentChildType, QueryBuilder<?> qb) {
+        this.parentChildType = parentChildType;
+        this.query = qb;
+        if (name == null) {
+            name = parentChildType;
+        }
+        inlineLeafInnerHits(qb);
+        return this;
+    }
+
     public InnerHitBuilder setNestedPath(String nestedPath) {
         this.nestedPath = nestedPath;
         return this;
     }
 
+    public InnerHitBuilder setNestedPath(String nestedPath, QueryBuilder<?> qb) {
+        this.nestedPath = nestedPath;
+        this.query = qb;
+        if (name == null) {
+            name = nestedPath;
+        }
+        inlineLeafInnerHits(qb);
+        return this;
+    }
+
+    // Inlines the inner hits of leaf queries with this inner hit,
+    // this is for hierarchical inner hits
+    void inlineLeafInnerHits(QueryBuilder<?> qb) {
+        if (qb instanceof HasChildQueryBuilder) {
+            HasChildQueryBuilder hcqb = (HasChildQueryBuilder) qb;
+            InnerHitBuilder ihb = hcqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                hcqb.clearInnerHit();
+            }
+        } else if (qb instanceof HasParentQueryBuilder) {
+            HasParentQueryBuilder hpqb = (HasParentQueryBuilder) qb;
+            InnerHitBuilder ihb = hpqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                hpqb.clearInnerHit();
+            }
+        } else if (qb instanceof NestedQueryBuilder) {
+            NestedQueryBuilder nqb = (NestedQueryBuilder) qb;
+            InnerHitBuilder ihb = nqb.innerHit();
+            if (ihb != null) {
+                addChildInnerHit(ihb);
+                nqb.clearInnerHit();
+            }
+        } else if (qb instanceof ConstantScoreQueryBuilder) {
```

true, this would make the result clearer, instead of silently not inlining the child inner hits and error would be thrown.